test/test_minmax_heap.jl

```diff
@@ -0,0 +1,151 @@
+    end
+    
+    @testset "pop! tests" begin
+        @testset "popmin! tests" begin
+            h = BinaryMinMaxHeap([1])
+            @test popmin!(h) == 1
+            @test length(h) == 0
+            @test isempty(h)
+            h = BinaryMinMaxHeap([1, 3, 2])
+            @test popmin!(h) == 1
+            @test length(h) == 2
+            @test is_minmax_heap(h.valtree)
+            @test minimum(h) == 2
+            for i = 1:20
+                A = rand(50)
+                h = BinaryMinMaxHeap(A)
+                minval = minimum(A)
+                @test popmin!(h) == minval
+                @test length(h) == 49
+                @test is_minmax_heap(h.valtree)
+            end
+        end
+        @testset "popmax! tests" begin
+            h = BinaryMinMaxHeap([1])
+            @test popmax!(h) == 1
+            @test length(h) == 0
+            @test isempty(h)
+            h = BinaryMinMaxHeap([1, 2, 3])
+            @test popmax!(h) == 3
+            @test length(h) == 2
+            @test is_minmax_heap(h.valtree)
+            @test maximum(h) == 2
+            h = BinaryMinMaxHeap([1, 3, 2])
+            @test popmax!(h) == 3
+            @test length(h) == 2
+            @test is_minmax_heap(h.valtree)
+            @test maximum(h) == 2
+        end
+    end
+    
+    @testset "empty!" begin
+        h = BinaryMinMaxHeap([1, 4, 3, 10, 2])
+        ret = empty!(h)
+        @test ret === h
+        @test length(ret) == 0
+        @test isempty(ret)
+    end
+
+    @testset "popall!" begin
+        @testset "popmin! tests" begin
+            A = rand(Int, 50)
+            sorted_A = sort(A)
+            h = BinaryMinMaxHeap(A)
+            @test popall!(h) == sorted_A
+            @test isempty(h)
+            @test length(h) == 0
+            
+            h = BinaryMinMaxHeap(A)
+            @test popmin!(h, 10) == sorted_A[1:10]
+            @test !isempty(h)
+            @test length(h) == 40
+        end
+        @testset "popmax! tests" begin
+            A = rand(Int, 50)
+            sorted_A = sort(A, order=Reverse)
+            h = BinaryMinMaxHeap(A)
+            @test popall!(h, Reverse) == sorted_A
+            @test length(h) == 0
+            @test isempty(h)
+            
+            h = BinaryMinMaxHeap(A)
+            @test popmax!(h, 10) == sorted_A[1:10]
+            @test length(h) == 40
+            @test !isempty(h)
+        end
+    end
+    
+    @testset "type conversion" begin
+        h = BinaryMinMaxHeap{Float64}()
+        push!(h, 3.)
+        push!(h, 5)
+        push!(h, Rational(4, 8))
+        push!(h, Complex(10.1, 0.0))
+        
+        @test h.valtree == [0.5, 10.1, 3.0, 5.0]
+    end
+    
+    @testset "descendants tests" begin
+        @test descendants(1, 1) == Int[]
+        @test descendants(2, 1) == [2]
+        @test descendants(3, 1) == [2, 3]
+        @test descendants(7, 1) == [2, 4, 5, 3, 6, 7]
+        @test descendants(10, 2) == [4, 8, 9, 5, 10]
+    end
+end
```

Comment:
Can you add tests to make sure the behavior of 
`pop!` etc on a empty heap throws
an appropriate error,
e.g.
```
ArgumentError: array must be non-empty
```

I think it does, but it might not.
And it is good to have these locked in to the test.

Showing it throws the errors that are expected

Reply:
It seems like returning 0 elements when asked for 1, is no more (nor less) incorrect than returning 1 element when asked for 2. etc.
As such I think returning an empty list is the most consistant behavour.

If different behavour was wanted, I think it should be around what to do if
`popmin!(h,k)` or `popmax(h, k)` are called with  `length(h) < k`.
Two option's there would be to

 - always error unless  unless a `allow_partial=true` kwarg is set?
 - Or conversely,  not error unless a `strict=true` kwarg is set.

Reply:
I can't think of a compelling reason to enforce stricter requirements (even as kwargs) from a user perspective, as long as it's clear in the docs that `length(popmin!(h, k)) <= k`.

The other reason one might care about this is if there are some nontrivial performance advantages to knowing the length of the array returned by `popmin!(h, k)`/`popmax!(h, k)`. I'm not in a position to say if there are any such advantages, however.

Unless you can think of anything further to add on those 2 points, I'm OK merging it as-is.